### DISCUSSION TITLE
Snapshot for Mac Catalyst [possible solution]

### DISCUSSION THREAD

**Discussion by huven** [fastlane/fastlane#21757] [Ideas]
It looks like _snapshot_ currently does not support Mac Catalyst. Research shows several others have tried this with little to no success, please reply if you did succeed!

After some testing I found that _snapshot_ uses the "~/Library/Caches/tools.fastlane" directory to communicate with the UI Tests target. And this causes the issue since the UI Tests Runner is sandboxed.

So, I made some changes which make the process work for me:

- changed SnapshotHelper.swift to set SIMULATOR_DEVICE_NAME env variable to 'Mac' for Mac Catalyst since the rest of the existing code assumes its existence
- changed SnapshotHelper.swift to use ~/Library/Caches/tools.fastlane as CacheDirectory
- since this is a directory outside the sandbox, I also added a .entitlements file to the UI Tests framework allowing access to this directory
- changed SnapshotHelper.swift to use the app window instead of the main screen to create the snapshot
- added some code to the app's scene delegate that resizes the main window to 1280x800 (if run within _snapshot_) to accommodate the App Store requirement

For me, this setup works, I now get perfect screenshots suitable for upload to the App Store.

Really interested in feedback from the community, is this the right way? Another approach is changing _snapshot_ to use a folder inside UI Tests sandbox, though I'm not sure how to predict its location reliably.

If there is enough support for the approach outlined above, I could turn it into a PR.

**Comment by donaldpiret** [comment]
A PR would be a lifesaver for me for this!

**Comment by phjs** [comment]
@huven thank you for pointing me to the entitlements. I did the same steps, but struggled to find a simple solution granting access to the directory outside the sandbox.

**Comment by huven** [comment]
You are welcome. I see I didn't attach the entitlements file, here is mine in case it helps anybody (uploaded as zip since .entitlements is not allowed by Github)
[UI Tests.entitlements.zip](https://github.com/user-attachments/files/16624345/UI.Tests.entitlements.zip)

**Comment by phjs** [comment]
For me, it only works when I add _App Sandbox_ with the value _NO_. Then the _com.apple.security.files.user-selected.read-write_ entry can be left empty.

@huven can you generate localized screenshots? My screenshots are all in system language, although the console prints `FASTLANE_LANGUAGE` is set to a different language for each run.

**Comment by huven** [comment]
Yup, localized snapshots work, that was the main reason I needed this. I do remember having an issue with that too, but if I remember correctly it was caused by the same issue. Not really sure though, but I'm pretty sure that I didn't need to make any code changes in the end for it, there was some secondary thing not ok.

**Comment by phjs** [comment]
I am still struggling to make it work. I know it's not the right place, but I don't know how else to contact you, so would you mind sharing how you configured fastlane to localize a catalyst app running on a Mac?

**Comment by jeffreyjackson** [comment]
A fix would be helpful. 👍🏻

**Comment by phjs** [comment]
After realizing that the language settings were somehow being ignored, I finally got it to work by adding the following code to the setUp method of my UI test.

```
let app = XCUIApplication()
#if targetEnvironment(macCatalyst)
    // Set the language and locale (again) to ensure the macOS simulator is localized (correctly).
    // ⚠️ Unfortunately, this doesn't ensure the right reading direction. Arabic UI is still in LTR
    // instead of RTL.
    let lang = Snapshot.deviceLanguage
    let locale: String = switch lang {
        case "de": "de_DE"
        case "ar": "ar_SA"
        case "en": "en_US"
        case "ru": "ru_RU"
        case "zh": "zh_Hans"
        default: "de_DE"
    }
    app.launchArguments += ["-AppleLanguages", "(\(lang))"]
    app.launchArguments += ["-AppleLocale", locale]
#endif
